Add ability to create PKey from parameters

huwcbjones opened the pull request:
Work towards #2047 by adding ability to create PKeys from OSSL params.

Also:
* tidy up creation of `OSSL_PARAM_*` on the way.
* write some tests for the internal OsslParam* API.

huwcbjones committed:
ossl_param: add tests

huwcbjones committed:
ossl_params/arrayref: refactor locate_octet_string to use locate

huwcbjones committed:
pkey_ctx: add Selection trait for Param/Public/Private markers

huwcbjones committed:
pkey_ctx: add ability to create a PKey from params

huwcbjones committed:
pkey_ctx/set_nonce_type: use ParamBuilder

Simplifies and safens the code

huwcbjones commented:
@alex, any chance you could take a look at this PR?

alex commented:
So, I'm not interested in including ossl params anywhere in the public API, I think they're a giant mistake and I hope OpenSSL gets rid of them.

I'd be ok:

- With a PR that just adds the -sys bindings
- Refactoring the private usage of OSSL_PARAMS to have a param builder or similar

But I'm not interested in any changes to the public APIs for this.

huwcbjones committed:
macros: add cstr_const

Let's minimise const cstr declaration boiler-plate

huwcbjones committed:
kdf: use cstr_const

huwcbjones committed:
kdf/argon2 helpers: make kdf names const

No point creating a new CStr every call

huwcbjones committed:
pkey: add common OSSL params

huwcbjones committed:
partial revert: Remove unused param builder methods

Partially reverts b7495f00c3bd9adcb117225875476ee218624cab by
adding back add_bn and add_utf_string

huwcbjones committed:
pkey: implement to_data

huwcbjones committed:
pkey: test round-robin to_data => from_data

huwcbjones committed:
ossl_param/array: implement merge

huwcbjones committed:
ossl_param/array: implement locate_bn

huwcbjones commented:
@alex apologies for the delay, had some higher priority stuff to look at. Please could you cast your eyes over this one?
(See #2464 for how these changes fit in)
